src/createEpicMiddleware.ts

```diff
@@ -106,8 +135,49 @@ export function createEpicMiddleware<
         'epicMiddleware.run(rootEpic) called before the middleware has been setup by redux. Provide the epicMiddleware instance to createStore() first.'
       );
     }
+
+    if (!rootEpic.id) {
+      Object.defineProperty(rootEpic, 'id', {
+        value: Math.random().toString(16),
+      });
+    }
+
     epic$.next(rootEpic);
   };
 
+  epicMiddleware.sigterm = (rootEpic) => {
+    const id = rootEpic.id;
+    const thread = threads[id];
+
+    if (!thread) {
+      // thread is not running,
+      // probably need to throw
+      return;
+    }
+
+    setTimeout(() => {
+      thread.sigterm$.next();
+    }, 0);
+
+    return thread.output$;
+  };
+
+  epicMiddleware.sigkill = (rootEpic) => {
+    const id = rootEpic.id;
+    const thread = threads[id];
+
+    if (!thread) {
+      // thread is not running,
+      // probably need to throw
+      return;
+    }
+
+    setTimeout(() => {
+      thread.sigkill$.next();
+    }, 0);
+
+    return thread.output$;
+  };
+
```

Comment:
### Preconditions

1. every running epic (thread) should be assigned a thread `id` to allow the middleware (thread manager) to terminate/kill threads individually

### Interface

1. **sigterm**: terminates the actions stream so that the thread would still have the time to complete any on-going job (e.g. async tasks), emit their last output and then exit gracefully
2. **sigkill**: terminates the thread directly, any on-going job would be abruptly killed straight away.

### Potential Usage

```typescript
const index = async (req, res) => {
  epicMiddleware.run(rootEpic);
  store.dispatch({ type: 'FETCH_TODOS' });
  
  const thread$ = epicMiddleware.sigterm(rootEpic);
  await thread$.toPromise();


  ReactDOMServer.renderToNodeStream(<Provider store={store}><App /></Provider>).pipe(res);
};
```

### Pros

1. Being able to manage every process individually (`run`, `kill`, `terminate`).

### Cons

1. if a rootEpic is ran multiple times, then `sigkill` and `sigterm` would affect all their instances (not necessarily a problem though)
2. `sigterm` can only terminate the action stream, so epics not propagating this completion would end up in a deadlock (as brilliantly highlighted by @liamzdenek [here](https://github.com/redux-observable/redux-observable/pull/539#issuecomment-543479927))

### Workarounds

1. Implement timeout to kill deadlocking processes

```typescript
const ensureShutdown = (middleware, rootEpic) => {
  setTimeout(() => {
    middleware.sigkill(rootEpic);
  }, 2000);
};

const index = async (req, res) => {
  epicMiddleware.run(rootEpic);
  store.dispatch({ type: 'FETCH_TODOS' });
  
  const thread$ = epicMiddleware.sigterm(rootEpic);
  ensureShutdown(epicMiddleware, rootEpic)
  await thread$.toPromise();


  ReactDOMServer.renderToNodeStream(<Provider store={store}><App /></Provider>).pipe(res);
};
```

### Additional (probably needed) steps

- `sigterm` should also send the signal to the `StateObservable`
- add methods to sigkill/sigterm all the running epics at once.

___

### Motivations

While SSR is probably one of the main use cases for implementing such API, it would also find applications in projects that heavily rely in code splitting. 

I am currently working on a project that mounts/unmounts reducers, components and epics depending on routing.

While we can easily add/remove reducers and components at runtime, it is impossible to do so with epics, since they cannot be terminated once ran.

**It really feels like `run` is missing its complementary api `terminate`**. (e.g. [redux#replaceReducer](https://redux.js.org/api/store#replacereducernextreducer))

### Should this feature live in RO or be implemented in a separate package?

It is my opinion that a complementary api for `run` is needed to reach feature completeness, nevertheless, shipping first-class support for SSR would drive adoption :)